README.md
# update_pipeline
Update and validate a pipelines metadata

[![Build Status](https://travis-ci.org/sanger-pathogens/update_pipeline.svg?branch=master)](https://travis-ci.org/sanger-pathogens/update_pipeline)   
[![License: GPL v3](https://img.shields.io/badge/License-GPL%20v3-brightgreen.svg)](https://github.com/sanger-pathogens/update_pipeline/blob/master/GPL-LICENSE)   

## Contents
  * [Introduction](#introduction)
  * [Installation](#installation)
    * [Required dependencies](#required-dependencies)
    * [Debian/Ubuntu](#debianubuntu)
    * [Running the tests](#running-the-tests)
  * [Usage](#usage)
    * [update\_pipeline\.pl](#update_pipelinepl)
    * [update\_pb\_pipeline](#update_pb_pipeline)
    * [update\_pipeline\_from\_spreadsheet\.pl](#update_pipeline_from_spreadsheetpl)
    * [validate\_pipeline\.pl](#validate_pipelinepl)
  * [License](#license)
  * [Feedback/Issues](#feedbackissues)

## Introduction
The scripts in this repository can be used to update and validate the metadata for a pipeline.

## Installation

update_pipeline has the following dependencies:

### Required dependencies
* [fastqcheck](https://github.com/sanger-pathogens/fastqcheck)
* [VR-Codebase](https://github.com/sanger-pathogens/vr-codebase.git)
* [MySQL](https://www.mysql.com/downloads/)

There are a few ways to install update-pipeline. If you encounter an issue when installing update_pipeline please contact your local system administrator. If you encounter a bug please log it [here](https://github.com/sanger-pathogens/update_pipeline/issues) or email us at path-help@sanger.ac.uk

### Debian/Ubuntu
These instructions assume you have root access and a working MySQL set up. To install update_pipeline and its dependencies, run:

```
apt-get update -qq
apt-get install -y libmysqlclient-dev autoconf libtool
git clone https://github.com/sanger-pathogens/update_pipeline.git && cd update_pipeline
./install_dependencies.sh
dzil authordeps --missing | cpanm
dzil listdeps --missing | cpanm
```
### Running the tests
The test can be run with dzil from the top level directory:  
  
`dzil test`  

## Usage
update_pipeline consists of four scripts for updating the metadata of a pipeline.

### update_pipeline.pl
```
Usage: ./update_pipeline.pl
  -s|--studies                 <file of study names>
  -n|--study_name              <a single study name to update>
  -d|--database                <vrtrack database name>
  -f|--max_files_to_return     <optional limit on num of file to check per process>
  -p|--parallel_processes      <optional number of processes to run in parallel, defaults to 1>
  -v|--verbose                 <print out debugging information>
  -r|--min_run_id              <optionally filter out errors below this run_id, defaults to 10000>
  -u|--update_if_changed       <optionally delete lane & file entries, if metadata changes, for reimport>
  -w|--dont_use_warehouse      <dont use the warehouse to fill in missing data>
  -tax|--taxon_id              <optionally provide taxon id to overwrite species info in bam file common name>
  -spe|--species                     <optionally provide the species name, which in combination with -tax avoids an NCBI lookup>
  -sup|--use_supplier_name     <optionally use the supplier name from the warehouse to populate name and hierarchy name of the individual table>
  -run|--specific_run_id       <optionally provide a specfic run id for a study>
  -min|--specific_min_run      <optionally provide a specfic minimum run id for a study to import>
  -nop|--no_pending_lanes      <optionally filter out lanes whose npg QC status is pending>
  -md5|--override_md5          <optionally update md5 on imported file if the iRODS md5 changes>
  -wdr|--withdraw_del          <optionally withdraw a lane if has been deleted from iRODS>
  -trd|--include_total_reads   <optionally write the total_reads from bam metadata to the file table in vrtrack>
  -l|--lock_file               <optional lock file to prevent multiple instances running>
  -t|--file_type               <optionally change the default file type to import from bam (to cram)>
  -h|--help                    <this message>

Update the tracking database from IRODs and the warehouse.

# update all studies listed in the file in the given database
./update_pipeline.pl -s my_study_file -d pathogen_abc_track

# update only the given study
./update_pipeline.pl -n "My Study" -d pathogen_abc_track

# Lookup all studies listed in the file, but only update the 500 latest files in IRODs
./update_pipeline.pl -s my_study_file -d pathogen_abc_track -f 500

# perform the update using 10 processes
./update_pipeline.pl -s my_study_file -d pathogen_abc_track -p 10

# import cram files
./update_pipeline.pl -s my_study_file -d pathogen_abc_track --file_type cram
```
### update_pb_pipeline
```
Usage: update_pb_pipeline [options]
Update the metadata of PB samples from the warehouse and iRODS and insert it into a VRTracking database.

  -s|--studies                 <file of study names>
  -n|--study_name              <a single study name to update>
  -d|--database                <vrtrack database name>
  -f|--max_files_to_return     <optional limit on num of file to check per process>
  -p|--parallel_processes      <optional number of processes to run in parallel, defaults to 1>
  -v|--verbose                 <print out debugging information>
  -r|--min_run_id              <optionally filter out errors below this run_id, defaults to 6000>
  -u|--update_if_changed       <optionally delete lane & file entries, if metadata changes, for reimport>
  -w|--dont_use_warehouse      <dont use the warehouse to fill in missing data>
  -tax|--taxon_id              <optionally provide taxon id to overwrite species info in bam file common name>
  -spe|--species               <optionally provide the species name, which in combination with -tax avoids an NCBI lookup>
  -sup|--use_supplier_name     <optionally use the supplier name from the warehouse to populate name and hierarchy name of the individual table>
  -run|--specific_run_id       <optionally provide a specfic run id for a study>
  -min|--specific_min_run      <optionally provide a specfic minimum run id for a study to import>
  -nop|--no_pending_lanes      <optionally filter out lanes whose npg QC status is pending>
  -md5|--override_md5          <optionally update md5 on imported file if the iRODS md5 changes>
  -wdr|--withdraw_del          <optionally withdraw a lane if has been deleted from iRODS>
  -trd|--include_total_reads   <optionally write the total_reads from bam metadata to the file table in vrtrack>
  -l|--lock_file               <optional lock file to prevent multiple instances running>
  -h|--help                    <this message>

# update all studies listed in the file in the given database
update_pb_pipeline -s my_study_file -d pathogen_abc_track

# update only the given study
update_pb_pipeline -n "My Study" -d pathogen_abc_track

# This help message
update_pb_pipeline -h
```
### update_pipeline_from_spreadsheet.pl
```
Usage: ./update_pipeline_from_spreadsheet.pl [options] spreadsheet.xls
  -d|--database                <vrtrack database name>
  -v|--verbose                 <print out debugging information>
  -f|--files_to_add_directory  <base directory containing files to add to the pipeline>
  -p|--pipeline_base_directory <required if -f provided, path to sequencing pipeline root directory>
  -u|--update_if_changed       <optionally delete lane & file entries, if metadata changes, for reimport>
  -t|--threads                 <number of threads to use>
  -r|--data_access_group       <restrict access to this unix group>

  -h|--help                    <this message>

# update the database only
./update_pipeline_from_spreadsheet.pl -d pathogen_abc_track spreadsheet.xls

# update the database and copy the sequencing files
./update_pipeline_from_spreadsheet.pl -d pathogen_abc_track -f /path/to/incoming/sequencing_files -p /lustre/scratch10x/xxx/seq-pipelines spreadsheet.xls
```
### validate_pipeline.pl
```
Usage: ./validate_pipeline.pl
            --studies               <study name or file of SequenceScape study names>
            [--database             <vrtrack database name>]
            --checkreadcount        <activate read count consistency evaluation (IO intensive)>
            -nop|--no_pending_lanes <optionally filter out lanes whose npg QC status is pending>
            --file_type             <cram or bam, defaults to bam>
            --specific_min_run      <dont check lanes below this run id (default 10000)>
            --help                  <this message>

Check to see if the pipeline is valid compared to the data stored in IRODS
```

## License
update_pipeline is free software, licensed under [GPLv3](https://github.com/sanger-pathogens/update_pipeline/blob/master/GPL-LICENSE).

## Feedback/Issues
Please report any issues to the [issues page](https://github.com/sanger-pathogens/update_pipeline/issues) or email path-help@sanger.ac.uk


CHANGELOG.md
# Change Log

## [Unreleased](https://github.com/sanger-pathogens/update_pipeline/tree/HEAD)

[Full Changelog](https://github.com/sanger-pathogens/update_pipeline/compare/v0.0.1...HEAD)

**Merged pull requests:**

- Include tests in README [\#107](https://github.com/sanger-pathogens/update_pipeline/pull/107) ([ssjunnebo](https://github.com/ssjunnebo))
- Update README [\#106](https://github.com/sanger-pathogens/update_pipeline/pull/106) ([ssjunnebo](https://github.com/ssjunnebo))
- 574328 fix travis tests [\#105](https://github.com/sanger-pathogens/update_pipeline/pull/105) ([ssjunnebo](https://github.com/ssjunnebo))
- remove version nr, autoversion instead [\#104](https://github.com/sanger-pathogens/update_pipeline/pull/104) ([ssjunnebo](https://github.com/ssjunnebo))

## [v0.0.1](https://github.com/sanger-pathogens/update_pipeline/tree/v0.0.1) (2018-01-25)
**Merged pull requests:**

- Add version number [\#103](https://github.com/sanger-pathogens/update_pipeline/pull/103) ([ssjunnebo](https://github.com/ssjunnebo))
- Fix Travis Errors [\#102](https://github.com/sanger-pathogens/update_pipeline/pull/102) ([ssjunnebo](https://github.com/ssjunnebo))
- change order of imeta query to speed things up [\#101](https://github.com/sanger-pathogens/update_pipeline/pull/101) ([andrewjpage](https://github.com/andrewjpage))
- add run\_date from mlwarehouse [\#100](https://github.com/sanger-pathogens/update_pipeline/pull/100) ([andrewjpage](https://github.com/andrewjpage))
- Get insert size from new warehouse [\#99](https://github.com/sanger-pathogens/update_pipeline/pull/99) ([andrewjpage](https://github.com/andrewjpage))
- add hash to check common name [\#98](https://github.com/sanger-pathogens/update_pipeline/pull/98) ([ssjunnebo](https://github.com/ssjunnebo))
- update email address [\#97](https://github.com/sanger-pathogens/update_pipeline/pull/97) ([ssjunnebo](https://github.com/ssjunnebo))
- dont import if there is no run id [\#96](https://github.com/sanger-pathogens/update_pipeline/pull/96) ([andrewjpage](https://github.com/andrewjpage))
- filter out old pb lanes [\#95](https://github.com/sanger-pathogens/update_pipeline/pull/95) ([andrewjpage](https://github.com/andrewjpage))
- inconsistent naming of library [\#94](https://github.com/sanger-pathogens/update_pipeline/pull/94) ([andrewjpage](https://github.com/andrewjpage))
- populate library name and sample public name if missing [\#93](https://github.com/sanger-pathogens/update_pipeline/pull/93) ([andrewjpage](https://github.com/andrewjpage))
- if library or sample missing fill in with placeholder data [\#92](https://github.com/sanger-pathogens/update_pipeline/pull/92) ([andrewjpage](https://github.com/andrewjpage))
- Pacbio duplicate libraries [\#91](https://github.com/sanger-pathogens/update_pipeline/pull/91) ([andrewjpage](https://github.com/andrewjpage))
- infgen [\#90](https://github.com/sanger-pathogens/update_pipeline/pull/90) ([andrewjpage](https://github.com/andrewjpage))
- update from 108 to 118 [\#89](https://github.com/sanger-pathogens/update_pipeline/pull/89) ([andrewjpage](https://github.com/andrewjpage))
- SRA fastq-dump header replacing [\#88](https://github.com/sanger-pathogens/update_pipeline/pull/88) ([vaofford](https://github.com/vaofford))
- add symlink rsync [\#87](https://github.com/sanger-pathogens/update_pipeline/pull/87) ([vaofford](https://github.com/vaofford))
- lookup xml directly first before using bioperl [\#86](https://github.com/sanger-pathogens/update_pipeline/pull/86) ([andrewjpage](https://github.com/andrewjpage))
- ncbi changed to https [\#85](https://github.com/sanger-pathogens/update_pipeline/pull/85) ([andrewjpage](https://github.com/andrewjpage))
- skip missing common name [\#84](https://github.com/sanger-pathogens/update_pipeline/pull/84) ([andrewjpage](https://github.com/andrewjpage))
- filter out lanes without study name [\#83](https://github.com/sanger-pathogens/update_pipeline/pull/83) ([andrewjpage](https://github.com/andrewjpage))
- pacbio should draw in results from irods [\#82](https://github.com/sanger-pathogens/update_pipeline/pull/82) ([andrewjpage](https://github.com/andrewjpage))
- nds\_change\_to\_dist\_zilla\_starter\_bundle [\#81](https://github.com/sanger-pathogens/update_pipeline/pull/81) ([nds](https://github.com/nds))
- bin directory should be set by softwarerc [\#80](https://github.com/sanger-pathogens/update_pipeline/pull/80) ([andrewjpage](https://github.com/andrewjpage))
- remove blank header cells [\#79](https://github.com/sanger-pathogens/update_pipeline/pull/79) ([andrewjpage](https://github.com/andrewjpage))
- Long running validate reports failing [\#78](https://github.com/sanger-pathogens/update_pipeline/pull/78) ([andrewjpage](https://github.com/andrewjpage))
- Fix for validate GCLP warehouse [\#77](https://github.com/sanger-pathogens/update_pipeline/pull/77) ([andrewjpage](https://github.com/andrewjpage))
- Data access group [\#76](https://github.com/sanger-pathogens/update_pipeline/pull/76) ([andrewjpage](https://github.com/andrewjpage))
- expose threads option for spreadsheet [\#75](https://github.com/sanger-pathogens/update_pipeline/pull/75) ([andrewjpage](https://github.com/andrewjpage))
- Run date [\#74](https://github.com/sanger-pathogens/update_pipeline/pull/74) ([andrewjpage](https://github.com/andrewjpage))
- update warehouse 2 cname [\#73](https://github.com/sanger-pathogens/update_pipeline/pull/73) ([andrewjpage](https://github.com/andrewjpage))
- only print sync message when verbose on [\#72](https://github.com/sanger-pathogens/update_pipeline/pull/72) ([andrewjpage](https://github.com/andrewjpage))
- Increase stability [\#71](https://github.com/sanger-pathogens/update_pipeline/pull/71) ([andrewjpage](https://github.com/andrewjpage))
- Reduce unnessisary lookups for warehouse fragment size [\#70](https://github.com/sanger-pathogens/update_pipeline/pull/70) ([andrewjpage](https://github.com/andrewjpage))

update_pipeline.pl
#!/usr/bin/env perl

=head1 NAME
=head1 SYNOPSIS
=head1 DESCRIPTION
=head1 CONTACT
=head1 METHODS

=cut

BEGIN { unshift(@INC, './modules') }
use strict;
use warnings;
no warnings 'uninitialized';
use POSIX;
use Getopt::Long;
use VRTrack::VRTrack;
use VRTrack::Lane;
use VertRes::Utils::VRTrackFactory;
use NCBI::SimpleLookup;
use NCBI::TaxonLookup;
use Parallel::ForkManager;

use UpdatePipeline::UpdateAllMetaData;
use UpdatePipeline::Studies;

my ( $studyfile, $help, $number_of_files_to_return, $lock_file, $parallel_processes, $verbose_output, $errors_min_run_id, $database,$input_study_name, $update_if_changed, $dont_use_warehouse, $taxon_id, $overwrite_common_name, $use_supplier_name, $specific_run_id, $specific_min_run, $common_name_required, $no_pending_lanes, $species_name, $override_md5, $withdraw_del, $vrtrack_lanes, $total_reads );

GetOptions(
    's|studies=s'               => \$studyfile,
    'n|study_name=s'            => \$input_study_name,
    'd|database=s'              => \$database,
    'f|max_files_to_return=s'   => \$number_of_files_to_return,
    'p|parallel_processes=s'    => \$parallel_processes,
    'v|verbose'                 => \$verbose_output,
    'r|min_run_id=s'            => \$errors_min_run_id,
    'u|update_if_changed'       => \$update_if_changed,
    'w|dont_use_warehouse'      => \$dont_use_warehouse,
    'tax|taxon_id=i'            => \$taxon_id,
    'spe|species=s'             => \$species_name,
    'sup|use_supplier_name'     => \$use_supplier_name,
    'run|specific_run_id=i'     => \$specific_run_id,
    'min|specific_min_run=i'     => \$specific_min_run,    
    'nop|no_pending_lanes'      => \$no_pending_lanes,
    'md5|override_md5'          => \$override_md5,
    'wdr|withdraw_del'          => \$withdraw_del,
    'trd|include_total_reads'   => \$total_reads,    
    'l|lock_file=s'             => \$lock_file,
    'h|help'                    => \$help,
);

my $db = $database ;

( ($studyfile || $input_study_name) &&  $db && !$help ) or die <<USAGE;
Usage: $0   
  -s|--studies                 <file of study names>
  -n|--study_name              <a single study name to update>
  -d|--database                <vrtrack database name>
  -f|--max_files_to_return     <optional limit on num of file to check per process>
  -p|--parallel_processes      <optional number of processes to run in parallel, defaults to 1>
  -v|--verbose                 <print out debugging information>
  -r|--min_run_id              <optionally filter out errors below this run_id, defaults to 6000>
  -u|--update_if_changed       <optionally delete lane & file entries, if metadata changes, for reimport>
  -w|--dont_use_warehouse      <dont use the warehouse to fill in missing data>
  -tax|--taxon_id              <optionally provide taxon id to overwrite species info in bam file common name>
  -spe|--species                     <optionally provide the species name, which in combination with -tax avoids an NCBI lookup>
  -sup|--use_supplier_name     <optionally use the supplier name from the warehouse to populate name and hierarchy name of the individual table>
  -run|--specific_run_id       <optionally provide a specfic run id for a study>
  -min|--specific_min_run      <optionally provide a specfic minimum run id for a study to import>  
  -nop|--no_pending_lanes      <optionally filter out lanes whose npg QC status is pending>
  -md5|--override_md5          <optionally update md5 on imported file if the iRODS md5 changes>
  -wdr|--withdraw_del          <optionally withdraw a lane if has been deleted from iRODS>
  -trd|--include_total_reads   <optionally write the total_reads from bam metadata to the file table in vrtrack>  
  -l|--lock_file               <optional lock file to prevent multiple instances running>
  -h|--help                    <this message>

Update the tracking database from IRODs and the warehouse.

# update all studies listed in the file in the given database
$0 -s my_study_file -d pathogen_abc_track

# update only the given study
$0 -n "My Study" -d pathogen_abc_track

# Lookup all studies listed in the file, but only update the 500 latest files in IRODs
$0 -s my_study_file -d pathogen_abc_track -f 500

# perform the update using 10 processes
$0 -s my_study_file -d pathogen_abc_track -p 10

USAGE

$parallel_processes ||= 1;
$verbose_output ||= 0;
$update_if_changed ||= 0;
$errors_min_run_id ||= 6000;
$dont_use_warehouse ||= 0;
$use_supplier_name ||=0;
$taxon_id ||= 0;
$common_name_required = $taxon_id ? 0 : 1;
$specific_run_id ||=0;
$specific_min_run ||=0;
$no_pending_lanes ||=0;
$override_md5 ||=0;
$withdraw_del ||=0;
$total_reads ||=0;

if(defined($lock_file))
{
  create_lock($lock_file);
}

my $study_names;

if(defined($studyfile))
{
  $study_names = UpdatePipeline::Studies->new(filename => $studyfile)->study_names;
}
else
{
  my @studyname = ($input_study_name);
  $study_names = \@studyname;
}

unless ($species_name) {
    eval{
      $species_name = $taxon_id ? NCBI::SimpleLookup->new( taxon_id => $taxon_id )->common_name : undef;
    };
    if ($@) {  
      eval {
    	$species_name = $taxon_id ? NCBI::TaxonLookup->new( taxon_id => $taxon_id )->common_name : undef;
      };
      die "Unable to retrieve taxonomic data from NCBI.\n" if ($@);	
    }
}

if($parallel_processes == 1)
{
  my $vrtrack = VertRes::Utils::VRTrackFactory->instantiate(database => $db,mode     => 'rw');
  unless ($vrtrack) { die "Can't connect to tracking database: $db \n";}
  if ( $withdraw_del ) {
    $vrtrack_lanes = {};
    foreach my $study (@$study_names) {
	  foreach my $lane ( $vrtrack->get_lanes(project => [$study]) ) {
	    $vrtrack_lanes->{ $lane->name } = $lane->id;
	  }
	}
  }   
  my $update_pipeline = UpdatePipeline::UpdateAllMetaData->new(
    study_names               => $study_names, 
    _vrtrack                  => $vrtrack, 
    number_of_files_to_return => $number_of_files_to_return, 
    verbose_output            => $verbose_output, 
    minimum_run_id            => $errors_min_run_id, 
    update_if_changed         => $update_if_changed,
    dont_use_warehouse        => $dont_use_warehouse,
    common_name_required      => $common_name_required,
    taxon_id                  => $taxon_id,
    species_name              => $species_name,
    use_supplier_name         => $use_supplier_name,
    specific_run_id           => $specific_run_id,
    specific_min_run          => $specific_min_run,
    no_pending_lanes          => $no_pending_lanes,
    override_md5              => $override_md5,
    vrtrack_lanes             => $vrtrack_lanes,
    add_raw_reads             => $total_reads,
  );
  $update_pipeline->update();
}
else
{
  my $pm = new Parallel::ForkManager($parallel_processes); 
  foreach my $study_name (@{$study_names}) {
    $pm->start and next; # do the fork
    my @split_study_names;
    push(@split_study_names, $study_name);
    my $vrtrack = VertRes::Utils::VRTrackFactory->instantiate(database => $db,mode     => 'rw');
    unless ($vrtrack) { die "Can't connect to tracking database: $db \n";}
    if ( $withdraw_del ) {
      foreach my $study (@$study_names) {
	    foreach my $lane ( $vrtrack->get_lanes(project => [$study]) ) {
		  $vrtrack_lanes->{ $lane->name } = $lane->id; 
		}
	  }
    }     
    my $update_pipeline = UpdatePipeline::UpdateAllMetaData->new(
      study_names               => \@split_study_names, 
      _vrtrack                  => $vrtrack, 
      number_of_files_to_return => $number_of_files_to_return, 
      verbose_output            => $verbose_output, 
      minimum_run_id            => $errors_min_run_id, 
      update_if_changed         => $update_if_changed,
      dont_use_warehouse        => $dont_use_warehouse,
      common_name_required      => $common_name_required,
      taxon_id                  => $taxon_id,
      species_name              => $species_name,
      use_supplier_name         => $use_supplier_name,
      specific_run_id           => $specific_run_id,
      specific_min_run          => $specific_min_run,
      no_pending_lanes          => $no_pending_lanes,
      override_md5              => $override_md5,
      vrtrack_lanes             => $vrtrack_lanes,
      add_raw_reads             => $total_reads,
    );
    $update_pipeline->update();
    $pm->finish; # do the exit in the child process
  }
  $pm->wait_all_children;
}

if(defined($lock_file))
{
  remove_lock($lock_file);
}


# Taken from vr-codebase/scripts/run-pipeline
sub create_lock
{
    my ($lock) = @_;
    if ( !$lock ) { return; } # the locking not requested

    if ( -e $lock )
    {
        # Find out the PID of the running pipeline
        my ($pid) = `cat $lock` || '';
        chomp($pid);
        if ( !($pid=~/^\d+$/) ) { print(qq[Broken lock file $lock? Expected number, found "$pid".\n]); }

        # Is it still running? (Will work only when both are running on the same host.)
        my ($running) = `ps h $pid`;
        if ( $running ) { die "Another process already running: $pid\n"; }
    }

    open(my $fh,'>',$lock) or die "Couldnt open lock file for writing";
    print $fh $$ . "\n";
    close($fh);

    return;
}

sub remove_lock
{
    my ($lock) = @_;
    if ( $lock && -e $lock ) { unlink($lock); }
    return;
}
